### PR TITLE
Implementation of Email Address International support (RFC 8398) in OpenSSL

### PR DESCRIPTION
**Pull request by beldmit**
RFC 8398 implementation in OpenSSL.

PUNYCODE decode is implemented from scratch.

Replaces #9199.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by beldmit** [comment]
Ping @openssl/committers?

**Comment by slontis** [comment]
Is there any existing test data out there in the wild? (The spec seems to be somewhat lacking in this department). It would be good to interop with something.

**Comment by beldmit** [comment]
> Is there any existing test data out there in the wild? (The spec seems to be somewhat lacking in this department). It would be good to interop with something.

AFAIK, no external test data. I have a set of certificates used for test purposes and plan to provide them via test scenario.

**Comment by beldmit** [comment]
Hopefully out of WIP.
Ping @openssl/committers

File: crypto/punycode.c
Comment (levitte) [inline crypto/punycode.c]: You need to add the usual copyright/license boilerplate...

Comment (beldmit) [reply crypto/punycode.c]: Done

**Comment by beldmit** [comment]
ping @openssl/committers after rebase.
@vdukhovni, could you please take a look?

**Comment by beldmit** [comment]
Force-pushed after #9333

**Comment by beldmit** [comment]
Ping @openssl/omc

File: 25-test_eai_data.t
Comment (levitte) [inline 25-test_eai_data.t]: This must be moved to `test/recipes/`

File: 25-test_eai_data/ascii_chain.pem
Comment (levitte) [inline 25-test_eai_data/ascii_chain.pem]: This must be moved to `test/recipes/25-test_eai_data/`

File: 25-test_eai_data/ascii_leaf.pem
Comment (levitte) [inline 25-test_eai_data/ascii_leaf.pem]: This as well

File: 25-test_eai_data/san.ascii
Comment (levitte) [inline 25-test_eai_data/san.ascii]: And this

File: 25-test_eai_data/san.utf8
Comment (levitte) [inline 25-test_eai_data/san.utf8]: And this

File: 25-test_eai_data/utf8_chain.pem
Comment (levitte) [inline 25-test_eai_data/utf8_chain.pem]: And this

File: 25-test_eai_data/utf8_leaf.pem
Comment (levitte) [inline 25-test_eai_data/utf8_leaf.pem]: And this

File: crypto/build.info
Comment (levitte) [inline crypto/build.info]: Wrong place. The providers aren't using this, so move it to the SOURCE defs of `../libcrypto`

File: crypto/punycode.c
Comment (levitte) [inline crypto/punycode.c]: Please rename this to `ossl_punicode_decode`

File: crypto/punycode.c
Comment (levitte) [inline crypto/punycode.c]: Please rename this to `ossl_a2ulabel`

File: crypto/punycode.c
Comment (levitte) [inline crypto/punycode.c]: Please rename this to `ossl_a2ucompare`

File: 25-test_eai_data.t
Comment (levitte) [inline 25-test_eai_data.t]: Year should be 2019, only

File: include/internal/punycode.h
Comment (levitte) [inline include/internal/punycode.h]: Copyright boilerplate missing.

Also, unless you think this is needed outside of libcrypto, please move this to `include/crypto/` and rename the `HEADER_INTERNAL_PUNYCODE_H` to something that's more like in the other files there.

File: 25-test_eai_data.t
Comment (beldmit) [reply 25-test_eai_data.t]: Fixed.

File: 25-test_eai_data/ascii_chain.pem
Comment (beldmit) [reply 25-test_eai_data/ascii_chain.pem]: Fixed.

File: 25-test_eai_data/ascii_leaf.pem
Comment (beldmit) [reply 25-test_eai_data/ascii_leaf.pem]: Fixed.

File: 25-test_eai_data/san.ascii
Comment (beldmit) [reply 25-test_eai_data/san.ascii]: Fixed.

File: 25-test_eai_data/san.utf8
Comment (beldmit) [reply 25-test_eai_data/san.utf8]: Fixed.

File: 25-test_eai_data/utf8_leaf.pem
Comment (beldmit) [reply 25-test_eai_data/utf8_leaf.pem]: Fixed.

File: crypto/build.info
Comment (beldmit) [reply crypto/build.info]: Done.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: include/internal/punycode.h
Comment (beldmit) [reply include/internal/punycode.h]: Done.

File: 25-test_eai_data.t
Comment (beldmit) [reply 25-test_eai_data.t]: Done.

**Comment by beldmit** [comment]
@levitte I hopefully implemented all the changes you've requested and made some git white magic to make the commits be logically consistent.

File: crypto/punycode.c
Comment (weihaw) [inline crypto/punycode.c]: Could a constant for 256 be used here? and below for bufsize? and align the constant in ossl_a2ucompare with this? i.e. justify 512 used there.  Can these limits be tested against the input string at the public interfaces?

File: crypto/x509/v3_ncons.c
Comment (weihaw) [inline crypto/x509/v3_ncons.c]: Apologies I'm not following the terminology for "`eml` and `base` hostname" . Perhaps local-part and domain part?

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: This module is an implementation of PUNYCODE conversion. Some parts of it are separated in helper procedures in RFC 3492, so I follow the guidelines from it.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Sorry, don't understand your point here :(

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Well. I did not find the information about how long are A-labels compared to U-labels. That's why I've limited it with maximum hostname length. It's worth (but seems not critical) to make them similar in size. 

It is possible to write some tests using these functions directly, but currently they are called via the name constraints validation procedure only.

File: crypto/x509/v3_ncons.c
Comment (beldmit) [reply crypto/x509/v3_ncons.c]: `eml` is the hostname from the email specified.
`base` is the hostname from the name constraint.

File: crypto/punycode.c
Comment (bernd-edlinger) [reply crypto/punycode.c]: I think he raises a valid point, that instead of using 256 at various places a define should be used.

**Comment by bernd-edlinger** [comment]
> This is my first time doing a github review let alone a openssl one.

Thanks @weihaw, you're welcome here!

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: The define is provided, thanks to @weihaw and @bernd-edlinger

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: @bernd-edlinger, is it OK now?

File: crypto/punycode.c
Comment (bernd-edlinger) [inline crypto/punycode.c]: I think you want `/*-` here

File: crypto/punycode.c
Comment (bernd-edlinger) [inline crypto/punycode.c]: indent this line by one space

File: crypto/punycode.c
Comment (bernd-edlinger) [inline crypto/punycode.c]: same here

File: crypto/punycode.c
Comment (bernd-edlinger) [inline crypto/punycode.c]: all comments...

File: crypto/punycode.c
Comment (bernd-edlinger) [inline crypto/punycode.c]: you'll probably want `/*-`
since the whitespace in the pseudocode below is significant.
You see?

File: crypto/punycode.c
Comment (bernd-edlinger) [inline crypto/punycode.c]: `*/` on a new line, please.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done

File: crypto/punycode.c
Comment (bernd-edlinger) [inline crypto/punycode.c]: pease just `*/`

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/x509/v3_ncons.c
Comment (bernd-edlinger) [inline crypto/x509/v3_ncons.c]: this line is too long

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/x509/v3_ncons.c
Comment (beldmit) [reply crypto/x509/v3_ncons.c]: Done.

File: crypto/punycode.c
Comment (bernd-edlinger) [inline crypto/punycode.c]: also too long, 80 column is maximum line length

File: crypto/punycode.c
Comment (bernd-edlinger) [inline crypto/punycode.c]: the space between `#` and define is only for ifdef blocks

File: crypto/punycode.c
Comment (bernd-edlinger) [inline crypto/punycode.c]: this does not line up with above.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Fixed

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Fixed

**Comment by beldmit** [comment]
Ping?

File: include/crypto/punycode.h
Comment (levitte) [inline include/crypto/punycode.h]: Documentation!  We're making a concerted effort to document our internal functions, so if you don't mind, please write some for these functions.  `doc/internal/man3/ossl_punycode_decode.pod`, say?

Comment (beldmit) [reply include/crypto/punycode.h]: Will do.

Comment (beldmit) [reply include/crypto/punycode.h]: Done.

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (levitte) [inline doc/internal/man3/ossl_punycode_decode.pod]: The indentation got a bit weird...

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (levitte) [inline doc/internal/man3/ossl_punycode_decode.pod]: Remove "function"

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (levitte) [inline doc/internal/man3/ossl_punycode_decode.pod]: Remove "an"

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (levitte) [inline doc/internal/man3/ossl_punycode_decode.pod]: Remove "function"

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (beldmit) [reply doc/internal/man3/ossl_punycode_decode.pod]: Fixed.

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (beldmit) [reply doc/internal/man3/ossl_punycode_decode.pod]: Done.

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (beldmit) [reply doc/internal/man3/ossl_punycode_decode.pod]: Done.

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (beldmit) [reply doc/internal/man3/ossl_punycode_decode.pod]: Done.

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (levitte) [inline doc/internal/man3/ossl_punycode_decode.pod]: These are arguments to `ossl_punycode_decode()`, but indented like this, they are too easy to mistake for a new function.  Please push it to the right to be aligned with the opening parenthesis.

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (levitte) [inline doc/internal/man3/ossl_punycode_decode.pod]: i misunderstood what this was trying to say, and went back to the function to see.  I suggest the following rephrase:

```
ossl_a2ucompare() accepts two NULL-terminated hostnames, decodes the 1st
from PUNYCODE to UTF-8 and compares it with the 2nd one as is.
```

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (levitte) [inline doc/internal/man3/ossl_punycode_decode.pod]: Suggested rephrase:

`ossl_punycode_decode() decodes one label (one dot-separated part) from a hostname,`

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (beldmit) [reply doc/internal/man3/ossl_punycode_decode.pod]: Many thanks! Done.

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (beldmit) [reply doc/internal/man3/ossl_punycode_decode.pod]: Done. Thank you very much for proofreading!

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (beldmit) [reply doc/internal/man3/ossl_punycode_decode.pod]: Done. Thank you very much for proofreading

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Done.

**Comment by beldmit** [comment]
Ping @vdukhovni @levitte 

BTW, I don't see the place where the changes are requested...

**Comment by mspncp** [comment]
> BTW, I don't see the place where the changes are requested...

FWIW: you can click [1] to get a list of reviewers and [2] to see the review.

image

It turns out that you marked all change requests already as resolved:

image


However, the red change request marker will only go away if @levitte re-reviews. But you have the permissions to officially re-request (or dismiss) the review yourself:

image

**Comment by beldmit** [comment]
Finally I got it. Many thanks!

**Comment by beldmit** [comment]
Ping @vdukhovni

**Comment by beldmit** [comment]
Ping @vdukhovni @levitte

**Comment by beldmit** [comment]
@vdukhovni, you told me there was a chance you'd look at it.

**Comment by beldmit** [comment]
Ping @openssl/otc.
All the changes requests seem to be addressed.

**Comment by romen** [comment]
@beldmit I see this PR has been waiting on action from @levitte or @vdukhovni for some time.
Unfortunately it also got into a conflict state: can you rebase it once more on master?

**Comment by beldmit** [comment]
> @beldmit I see this PR has been waiting on action from @levitte or @vdukhovni for some time.
> Unfortunately it also got into a conflict state: can you rebase it once more on master?

Done

File: crypto/punycode.c
Comment (levitte) [reply crypto/punycode.c:64]: Nope

File: crypto/punycode.c
Comment (levitte) [reply crypto/punycode.c:241]: Nope

File: crypto/punycode.c
Comment (levitte) [reply crypto/punycode.c:250]: Nope

File: crypto/punycode.c
Comment (levitte) [reply crypto/punycode.c:320]: Nope

File: test/recipes/25-test_eai_data.t
Comment (levitte) [inline test/recipes/25-test_eai_data.t:22]: I would make this  `#./util/wrap.pl apps/openssl ...`

**Comment by levitte** [comment]
Sorry, I was distracted elsewhere.  Apart from some remaining nits, this looks fine.

**Comment by beldmit** [comment]
@levitte, I'm afraid I created a merge commit by accident
How do I get rid of it?

File: test/recipes/25-test_eai_data.t
Comment (beldmit) [reply test/recipes/25-test_eai_data.t]: Fixed. Thanks!

**Comment by beldmit** [comment]
```git rebase --skip``` did the trick!

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Now done :)
Thanks!

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Now done :)
Thanks!

File: crypto/punycode.c
Comment (beldmit) [reply crypto/punycode.c]: Now done :)
Thanks!

**Review by levitte — APPROVED** [review]

**Comment by beldmit** [comment]
Squashed and merged. Many thanks!

**Comment by henmaster** [comment]
This didn't age well.

**Pull request closed without merging**
